Manage organization "Teams"

divbzero opened the pull request:
- Closes #11073.
- Closes #11074.
- Closes #11075.
- Closes #11079.
- Closes #11080.
- Closes #11082.
- Closes #11084.

ewdurbin commented:
> @ewdurbin When I was adding journal entries/events, I came across a few inconsistencies in action/tag names that I tried to reconcile with commit [2264221](https://github.com/pypa/warehouse/pull/11665/commits/226422180cd0c471736f909591fb565d13905804). Please let me know —
> 
> 1. Do those changes make sense?
> 2. If yes, should I remove that commit from this PR and create a separate PR with those changes?

A separate PR is probably best.

ewdurbin commented on warehouse/organizations/models.py:
```diff
@@ -311,3 +341,140 @@ class OrganizationInvitation(db.Model):
+            raise KeyError from None
+
+
+# TODO: Do we enable HasEvents functionality? Or are these just Org Events?
```
Since membership in teams can grant access, yes these should `HasEvents`

divbzero replied:
Enabled `HasEvents` in commit [c7e2dcc](https://github.com/pypi/warehouse/pull/11665/commits/c7e2dcc84a6ea11e4947dba6fa929254b9037443).

ewdurbin commented on warehouse/organizations/models.py:
```diff
@@ -311,3 +341,140 @@ class OrganizationInvitation(db.Model):
+    )
+
+    organization = orm.relationship("Organization", lazy=False, backref="teams")
+    # TODO: Or do we call this "members" instead?
```
members is probably better in my opinion.

divbzero replied:
Renamed to `members` in commit [11733f6](https://github.com/pypi/warehouse/pull/11665/commits/11733f688a8839adcb6dc39a4183a481c78b9cf5).

ewdurbin commented on warehouse/templates/manage/organization/teams.html:
```diff
@@ -0,0 +1,100 @@
+            {% endif %}
+          </h3>
+          <p class="package-snippet__meta">
+            {% trans member_count=team.members|length %}
```
This seems to always be calculating as zero. I think it should be `team.users` unless the name is changed on the model to `members` as mentioned in `TODO`

divbzero replied:
Renamed to `members` as mentioned above.

ewdurbin reviewed (commented):
A few minor things caught in review and some comments on TODO.

Overall this functions as expected. I believe user-testing will expose some specific user experience concerns, but I can't predict them at this time.

divbzero commented:
> A separate PR is probably best.

Created separate PR #11779 to update project role journal entries and events.

ewdurbin commented on warehouse/manage/forms.py:
```diff
@@ -36,7 +40,17 @@ class RoleNameMixin:
     role_name = wtforms.SelectField(
         "Select role",
-        choices=[("", "Select role"), ("Maintainer", "Maintainer"), ("Owner", "Owner")],
+        choices=[("", "Select role"), ("Maintainer", "Upload"), ("Owner", "Admin")],
```
We should leave the labels for existing non-org projects as `Maintainer` and `Owner` until further notice, as merging this would change these labels and cause confusion for users.

divbzero replied:
Fixed with commit [26bbc2e](https://github.com/pypi/warehouse/pull/11665/commits/26bbc2e824ce689cb4281287d0bec3e9dfb0b3fc).

ewdurbin reviewed (changes requested):
The only outstanding change that needs to be addressed before this is mergable is that we should not change the current "Owner" and "Maintainer" role names for non-org projects.

Not only due to confusion in the UI, but also due to existing documentation, our own FAQ/help page, and 3rd party docs that reference these names.

ewdurbin commented:
@divbzero note that this branch was rebased off of main to resolve translations merge conflicts that weren't auto-resolvable by GitHub

divbzero commented:
@ewdurbin I’ve reverted the roles for individual users back to `Maintainer` and `Owner`, renamed permissions from `Admin` to `Administer`, and updated the project Collaborators page to clarify the distinction between roles and permissions.

divbzero commented:
Fixed a few things that would have come up during user testing:
- Tweaked navigation for organizations and teams (commits [1c00802](https://github.com/pypi/warehouse/pull/11665/commits/1c00802f910f84257b7ed1833d2dc871422edecb) and [61e41b8](https://github.com/pypi/warehouse/pull/11665/commits/61e41b89bf8371310aa580fa0fa0cdf8b4aa85a1)).
- Fixed issue where teams with `Administer` permissions could’t actually administer the project (commit [3bce85c](https://github.com/pypi/warehouse/pull/11665/commits/3bce85c8f6fc364c1b41cbe0d84b524708a2a71f)).

Things should be cleaner now so users can focus on other usability issues.

@ewdurbin I think this pull request should be all set for merging. Please let me know if you spot anything else.

ewdurbin reviewed (approved):
This is working as expected and is ready for user testing.
